src/cli/tree.rs

```diff
@@ -0,0 +1,401 @@
+        let last = index == needed_count - 1;
+        let symbol = if last {
+            UTF8_SYMBOLS.ell
+        } else {
+            UTF8_SYMBOLS.tee
+        };
+
+        if let Some(needed_pkg) = inverted_dep_map.get(needed_name) {
+            print_package(
+                format!("{prefix}{symbol} "),
+                needed_pkg,
+                direct_deps.contains(&needed_pkg.name),
+                false,
+            );
+
+            let new_prefix = if index == needed_count - 1 {
+                format!("{}{} ", prefix, UTF8_SYMBOLS.empty)
+            } else {
+                format!("{}{} ", prefix, UTF8_SYMBOLS.down)
+            };
+
+            print_inverted_leaf(needed_pkg, new_prefix, inverted_dep_map, direct_deps)
+        }
+    }
+}
+
+/// Filter and print a top down dependency tree
+fn print_dependency_tree(
+    dep_map: &HashMap<String, Package>,
+    direct_deps: &Vec<String>,
+    regex: &Option<String>,
+) -> Result<(), miette::Error> {
+    let mut filtered_deps = direct_deps.clone();
+
+    if let Some(regex) = regex {
+        let regex = regex::Regex::new(regex).map_err(|_| miette::miette!("Invalid regex"))?;
+        filtered_deps.retain(|p| regex.is_match(p));
+
+        if filtered_deps.is_empty() {
+            Err(miette::miette!(
+                "No top level dependencies matched the given regular expression"
+            ))?;
+        }
+    }
+
+    let mut visited_pkgs = Vec::new();
+    let direct_dep_count = filtered_deps.len();
+
+    for (index, pkg_name) in filtered_deps.iter().enumerate() {
+        visited_pkgs.push(pkg_name.to_owned());
+
+        let last = index == direct_dep_count - 1;
+        let symbol = if last {
+            UTF8_SYMBOLS.ell
+        } else {
+            UTF8_SYMBOLS.tee
+        };
+        if let Some(pkg) = dep_map.get(pkg_name) {
+            print_package(format!("{symbol} "), pkg, true, false);
+
+            let prefix = if last {
+                UTF8_SYMBOLS.empty
+            } else {
+                UTF8_SYMBOLS.down
+            };
+            print_dependency_leaf(
+                pkg,
+                format!("{} ", prefix),
+                dep_map,
+                &mut visited_pkgs,
+                direct_deps,
+            )
+        }
+    }
+    Ok(())
+}
+
+/// Recursively print top down dependency tree nodes
+fn print_dependency_leaf(
+    pkg: &Package,
+    prefix: String,
+    dep_map: &HashMap<String, Package>,
+    visited_pkgs: &mut Vec<String>,
+    direct_deps: &Vec<String>,
+) {
+    let dep_count = pkg.dependencies.len();
+    for (index, dep_name) in pkg.dependencies.iter().enumerate() {
+        let last = index == dep_count - 1;
+        let symbol = if last {
+            UTF8_SYMBOLS.ell
+        } else {
+            UTF8_SYMBOLS.tee
+        };
+
+        // skip virtual packages
+        if dep_name.starts_with("__") {
+            continue;
+        }
```

Comment:
I can imagine this being useful 🤷

Reply:
I was having failures with the environments I tried, and as far as I could tell the virtual packages for the system were the ones with `__`. I'm guessing it's because they aren't a fully featured package in the environment.

Maybe I could add an `else` block to the `if let` on 250 to print them.